code/modules/projectiles/projectile.dm

```diff
@@ -435,89 +400,72 @@
  * Furthermore, this proc shouldn't check can_hit_target - this should only be called if can hit target is already checked.
  * Also, we select_target to find what to process_hit first.
  */
-/obj/projectile/proc/Impact(atom/A)
-	if(!trajectory)
-		qdel(src)
-		return FALSE
-	if(impacted[A.weak_reference]) // NEVER doublehit
-		return FALSE
-	var/datum/point/point_cache = trajectory.copy_to()
-	var/turf/T = get_turf(A)
-	if(ricochets < ricochets_max && check_ricochet_flag(A) && check_ricochet(A))
-		ricochets++
-		if(A.handle_ricochet(src))
-			on_ricochet(A)
-			impacted = list() // Shoot a x-ray laser at a pair of mirrors I dare you
-			ignore_source_check = TRUE // Firer is no longer immune
-			decayedRange = max(0, decayedRange - reflect_range_decrease)
-			ricochet_chance *= ricochet_decay_chance
-			damage *= ricochet_decay_damage
-			stamina *= ricochet_decay_damage
-			range = decayedRange
-			if(hitscan)
-				store_hitscan_collision(point_cache)
-			return TRUE
+/obj/projectile/proc/impact(atom/target)
+	if(impacted[target.weak_reference]) // never doublehit, otherwise someone may end up running into a projectile from the back
+		return
 
-	var/distance = get_dist(T, starting) // Get the distance between the turf shot from and the mob we hit and use that for the calculations.
-	def_zone = ran_zone(def_zone, clamp(accurate_range - (accuracy_falloff * distance), 5, 100)) //Lower accurancy/longer range tradeoff. 7 is a balanced number to use.
+	if(ricochets < ricochets_max && check_ricochet_flag(target) && check_ricochet(target) && target.handle_ricochet(src))
+		on_ricochet(target)
+		impacted = list() // Shoot a x-ray laser at a pair of mirrors I dare you
+		ignore_source_check = TRUE // Firer is no longer immune
+		decayedRange = max(0, decayedRange - reflect_range_decrease)
+		ricochet_chance *= ricochet_decay_chance
+		damage *= ricochet_decay_damage
+		stamina *= ricochet_decay_damage
+		range = decayedRange
+		if(hitscan && trajectory)
+			store_hitscan_collision(trajectory.copy_to())
+		return
 
-	return process_hit(T, select_target(T, A, A), A) // SELECT TARGET FIRST!
+	var/turf/target_turf = get_turf(target)
+	// Lower accurancy/longer range tradeoff. 7 is a balanced number to use.
+	def_zone = ran_zone(def_zone, clamp(accurate_range - (accuracy_falloff * get_dist(target_turf, starting)), 5, 100))
+	target = select_target(target_turf, target)
+	process_hit_loop(target)
+
+/obj/projectile/proc/process_hit_loop(atom/target)
```

Comment:
Deserves to be documented? 

Also maybe mark as should not override or something?